File: exercise/102/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) return result;
        queue<pair<TreeNode*, int>> fifo;
        fifo.push(make_pair(root, 0));
        int level = 0;
        vector<int> level_result;
        while (!fifo.empty()) {
            pair<TreeNode*, int> now = fifo.front();
            fifo.pop();
            TreeNode* node = now.first;
            if (now.second != level) {
                result.push_back(level_result);
                level_result.clear();
                level++;
            }
            level_result.push_back(node->val);
            if (node->left)
                fifo.push(make_pair(node->left, now.second + 1));
            if (node->right)
                fifo.push(make_pair(node->right, now.second + 1));
        }
        result.push_back(level_result);
        return result;
    }
};

File: exercise/102/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        from collections import deque
        result = []
        queue = deque()
        queue.append((root, 0))
        now_level = 0
        now = []
        while len(queue) > 0:
            node = queue.popleft()
            if node[0]:
                if node[1] > now_level:
                    result.append(now)
                    now = []
                now.append(node[0].val)
                queue.append(node[0].left)
                queue.append(node[0].right)
        if len(now) > 0:
            result.append(now)
        return result
